app/controllers/fhir/r4/api_controller.rb

```diff
@@ -102,53 +102,71 @@ def update
         :'system/Patient.*'
       )
 
-      resource = get_patient(params.permit(:id)[:id])
+      # Get the patient that needs to be updated
+      patient = get_patient(params.permit(:id)[:id])
+      status_forbidden && return if patient.nil?
 
-      if request.patch? && !resource.nil?
+      # Get the contents from applying a patch, if needed
+      if request.patch? && !patient.nil?
         begin
-          contents = apply_patch(resource, patch)
+          contents = apply_patch(patient, patch)
         rescue StandardError => e
           status_bad_request([['Unable to apply patch', e&.message].compact.join(': ')]) && return
         end
       end
 
-      updates = patient_from_fhir(contents, default_patient_jurisdiction_id)
+      # Get patient values before updates occur for later comparison
+      patient_before = patient.dup
 
-      # Grab patient before changes to construct diff
-      patient_before = resource.dup
+      # Get key value pairs from the update.
+      request_updates = patient_from_fhir(contents, default_patient_jurisdiction_id)
+      status_unprocessable_entity && return if request_updates.nil?
 
-      # If "monitoring" was set to false for a Patient that was previously being monitored
-      if !resource.nil? && resource.monitoring && updates&.key?(:monitoring) && !updates[:monitoring]
-        # Add closed_at to updates
-        updates[:closed_at] = DateTime.now
+      # Assign any remaining updates to the patient
+      # NOTE: The patient.update method does not allow a context to be passed, so first we assign the updates, then save
+      patient.assign_attributes(request_updates)
+      # Verify that the updated jurisdiction and other updates are valid
+      unless jurisdiction_valid_for_update?(patient) && patient.save(context: :api)
+        status_unprocessable_entity(format_model_validation_errors(patient)) && return
       end
-    else
-      status_not_found && return
-    end
-
-    status_forbidden && return if resource.nil?
 
-    # Try to update the resource
-    status_unprocessable_entity && return if updates.nil?
+      # If the jurisdiction was changed, create a Transfer
+      if request_updates&.keys&.include?(:jurisdiction_id) && !request_updates[:jurisdiction_id].nil?
+        Transfer.create(patient: patient, from_jurisdiction: patient_before.jurisdiction, to_jurisdiction: patient.jurisdiction, who: @current_actor)
```

Comment:
As @kylietmo commented in `PatientsController` I think we'd also want this to be a `create!` so we get an error if this fails for some reason.

On a somewhat related, I noticed that here the `Transfer` is created after saving the change on the patient, while in `PatientsController.udpate` it's created before saving the change on the patient. In the happy path I don't think the order would matter (though correct me if I'm wrong). But in both cases if something goes wrong with the _second_ save, the first save will still have taken affect. Unless I'm unaware of some rails magic going on (always possible), I think we may want to use [`Transactions`](https://api.rubyonrails.org/v6.1.0/classes/ActiveRecord/Transactions/ClassMethods.html) to make the two DB calls a single atomic action with rollback.

Reply:
Wow Transactions seem very useful for our situation, awesome. In https://github.com/SaraAlert/SaraAlert/pull/664/commits/b7b848846d6239155a5ab47612c59125fdb88aea, I updated the `APIController` to use a transaction so that updates, creation of transfers, and creation of history items is all atomic.

I was more hesitant to do this in the `PatientsController`, because in my estimation, adding transactions to the `PatientsController` would be pretty significant task to ensure we do it right. Not just `update`, but `update_status`, `update_hoh`, and `bulk_update` all have side effects (History item creation for example) that could probably happen atomically with updates. But as of right now, we don't even really have error handling in the `PatientsController`, if an `update` is done with invalid values, the user doesn't even get alerted at all. The monitoree appears to save, but the value does not actually change. I think generally it seems like we are trusting that what comes in from the front end won't cause an issue.

So I think adding transactions and more error handling to the `PatientsController` would require more time. I'm happy to do it as a part of this PR if we want, but I do think it would increase the PR size pretty significantly, and take a bit of time to do.

Reply:
Sound reasoning to hold off on the `PatientsController` for now. I think we'll also want a transaction in `ApiController.create` (for the patient and the history item), and possibly in other non-API controllers as well, but I'm good with making those future tasks.

I just created [SARAALERT-1236](https://tracker.codev.mitre.org/browse/SARAALERT-1236) and [SARAALERT-1237](https://tracker.codev.mitre.org/browse/SARAALERT-1237) as a starting point. @kylietmo do you agree with this?

Reply:
Yes, good call. Thanks all!